probtorch/stochastic.py

```diff
@@ -321,6 +324,30 @@ def log_joint(self, sample_dim=None, batch_dim=None, nodes=None):
                 log_prob = log_prob + log_p
         return log_prob
 
+    def log_pair(self, sample_dim, batch_dim, nodes):
+        """Return the log average probability values, optionally for a subset of nodes.
+
+        Arguments:
+            nodes: The subset of nodes for which the log pair probabilities to be computed.
+            sample_dim(int): The dimension that enumerates samples.
+            batch_dim(int): The dimension that enumerates batch items.
+        """
+
+        log_avg_joint = 0.0
+        log_avg_sep = 0.0
+        # Note: not entirely sure if this is correct for more than one node
+        for n in nodes:
+            if n in self._nodes:
+                node = self._nodes[n]
+                value = node.value
+                size = value.size()
+                values_expand = value.unsqueeze(3).repeat(1, 1, 1, size[batch_dim]).permute(batch_dim, sample_dim, 3, 2)
+                log_pair_probs = node.dist.log_prob(values_expand).permute(1, 0, 2, 3) # (SxBxBxD) dimension
+                log_avg_joint = log_avg_joint + log_mean_exp(log_pair_probs.sum(-1), 2)
+                log_avg_sep = log_avg_sep + log_mean_exp(log_pair_probs, 2).sum(-1)
```

Comment:
These lines need some more comments – I'm not sure I understand what is going on here.